src/hostname_resolution.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Library containing logic pertaining to hostname resolutions in the VM charm."""
+
+import io
+import logging
+import socket
+
+from ops.charm import CharmBase
+from ops.framework import Object
+from ops.model import Unit
+
+from constants import HOSTNAME_TO_IP_KEY, PEER
+
+logger = logging.getLogger(__name__)
+
+
+class MySQLMachineHostnameResolution(Object):
+    """Encapsulation of the the machine hostname resolution."""
+
+    def __init__(self, charm: CharmBase):
+        super().__init__(charm, "hostname-resolution")
+
+        self.charm = charm
+
+        self.framework.observe(self.charm.on.config_changed, self._update_hostname_ip_in_databag)
```

Comment:
When the IP changes, it takes some time for the `config-changed` event to fire (as much as it takes juju to update the IP address in the controller). However, the alternative for implementing an observer that constantly checks the unit's IP address and updates the peer databag is also an option (albeit a slightly more complex one)

Reply:
- Actually, the reason we use `socket` is because when the `config-changed` event is fired, the `instance-hostname` and `private-address` fields are not populated in the local unit databag.

- Empirically, it takes a few minutes to fire the `config-changed`. I think that the IP address change detection is tied to the `update-status` hook interval

Reply:
implemented an IP address observer to speed up discovery of changed addresses in d687dbc0aee31c83ce90877d7505006123359fe1